### PR TITLE
Setting up container goals

### PR DESCRIPTION
**Pull request by gjtorikian**
This has been a long under construction doc that I just finalized to match Reality :tm: with the recent release of [container goals](https://blog.atomist.com/announcing-sdm-1-6/). It should be considered ready for review.

Apologies on the whitespace changes in _mkdocs.yml_, VSCode's linter is opinionated. I can revert those changes if it's unnecessary, I don't mind them either way.

### CODE REVIEW

File: docs/developer/container-goals.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+1.  We'll build a JAR artifact from a Java project via a Docker image from the [Docker Image Registry](https://www.docker.com/products/image-registry); then
+2.  We'll feed that JAR into [Kaniko](https://github.com/GoogleContainerTools/kaniko) in order to build our application's container image.
+
+## Prerequisite
+
+You should have a general idea of both how the SDM operates, and the steps involved in building a Docker image. You'll also need to make sure that your local terminal has access to a Docker daemon; for example, if you have [minikube](https://github.com/kubernetes/minikube) installed, you can try running:
+
+```
+eval $(minikube docker-env)
```
Comment (jrday) [inline docs/developer/container-goals.md]: Could be interpreted to require a k8s? This only needs a docker daemon running, correct?

File: docs/developer/container-goals.md [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+          }],
+        }),
+      ],
+      test: ...,
+      dependsOn: ...
+    }
+  };
+```
+
+`reg`refers to the container goal registrations which are defined as the second argument to the`container`function, while`proj`is the project the SDM is running against. From the`proj`, it's possible to get the name of the underlying repository, which we can use to define a more unique resulting image, based on the SHA of the last pushed commit.
```
[the remark this answers is not in the document]
Comment (ddgenome) [reply docs/developer/container-goals.md]: A space is missing between "the`container`".

[the remark this answers is not in the document]
Comment (ddgenome) [reply docs/developer/container-goals.md]: A space is missing between "while`proj`".

[the remark this answers is not in the document]
Comment (ddgenome) [reply docs/developer/container-goals.md]: And "the`proj`"

File: docs/developer/container-goals.md [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+### Dynamic destination
+
+You'll notice that the destination of the resulting Docker image is a static location, which is less than ideal, because it means that you can't (easily) store multiple versions of the images. To alleviate this, container goals also have a `callback` property, which are executed right before the container goal. You can use this function to fetch information about the project that the SDM is operating within:
+
+```typescript
+  return {
+    jvm: { ... },
+    docker: {
+      goals: [
+        callback: async (reg, proj) => {
```
Comment (ddgenome) [inline docs/developer/container-goals.md]: I've updated the blog post with example code that may be more clear.

File: docs/developer/container-goals.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+      ],
+      test: ...,
+      dependsOn: ...
+    }
+  };
+```
+
+`reg`refers to the container goal registrations which are defined as the second argument to the`container`function, while`proj`is the project the SDM is running against. From the`proj`, it's possible to get the name of the underlying repository, which we can use to define a more unique resulting image, based on the SHA of the last pushed commit.
+
+## Complete sample
```
Comment (ddgenome) [inline docs/developer/container-goals.md]: Maybe this should point to a sample in the samples repo so it is more likely to stay working?

**Comment by gjtorikian** [comment]
> Sorry for all the comments.

Not at all! The more the better.

File: docs/developer/container-goals.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+```
+
+`goals` is an array of the actions to perform. In this case, we're only building our JAR, but we could perform anything else here, such as code autolinting.
+
+A container goal consists of a function, `container`, that takes two arguments. The first is a string, `build-jar`, and it behaves similarly to the [`displayName` for any Build action](https://docs.atomist.com/developer/build/#the-build-goal): it provides a more specific identifier for logs and output.
+
+After that comes the `containers` argument, which is where a bulk of the logic resides. A container goal executes a [`docker run`](https://docs.docker.com/engine/reference/commandline/run/) command, which is roughly format of the resulting command is in the form of
+
+```
+docker run [OPTIONS] IMAGE [COMMAND] [ARG...]
```
[the remark this answers is not in the document]
Comment (gjtorikian) [reply docs/developer/container-goals.md]: Nice!

File: docs/developer/container-goals.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjtorikian) [reply docs/developer/container-goals.md]: I sincerely have no idea how this happened. 😂

**Comment by gjtorikian** [comment]
Thanks for the review @ddgenome, I really appreciate it, as most of the content was gleaned from early access to the tooling and the blog post. There's only two changes I felt strongly enough about to disagree with accepting:

* I understand that `mkdocs` might prefer the `.md` extension for its own link resolution, but I don't think that that's safe enough to rely on. For example, perhaps a subsequent update will remove that functionality, or maybe the blog will move to a different platform altogether. I like using the relative URL path because the functionality is less surprising. 
* Regarding [`callback`](https://github.com/atomist/docs/pull/327#discussion_r304146370), I actually prefer how it was before, where a single argument is pushed rather then entire object being constructed at once. Is there a preference for one over the other?

**Comment by ddgenome** [comment]
-   Fair enough. My reasons were two-fold. First, it was quicker to find those link errors. Second, it turns out the HTML checker was not working for a while, so we had no link-check fallback.
-   Either way should work and you should use whatever you think is clearer. One thing to note is that the `callback` does not return the entire goal registration, although it can. From the `callback` return value, only `containers` and `volumes` are used. The other parts of the registration cannot be changed during execution.

**Comment by gjtorikian** [comment]
> One thing to note is that the callback does not return the entire goal registration, although it can.

Ah, good to note!

**Comment by gjtorikian** [comment]
Just a head's up that I don't think the source sample SDM checks broken links -- or at least, all tests passed here, but I had to fix a reference in 7179228.

Anywho -- I think this is good to merge, pending someone's ✅ .

**Comment by ddgenome** [comment]
@gjtorikian , If you look at the full message for the build here https://atomist-community.slack.com/archives/C3A8JLX7Y/p1563472711012000 , you'll see that htmltest did indeed fail because of that broken link.

**Comment by gjtorikian** [comment]
Oh, huh. I was just going off the GitHub Status. 🤷🏻‍♂️

**Comment by ddgenome** [comment]
@jessitron did not want htmltest failures to result in GitHub status check failures.

**Comment by jessitron** [comment]
hmm I could probably allow htmltest to fail the gh status now that it excludes failures due to files not yet existing on master.

**Pull request merged**